### ISSUE TITLE
miniwdl-lsf is lacking support for runtime (`-W`) or queue (`-q`) arguments

### ISSUE THREAD

**Issue by Marc-Ruebsam** [adthrasher/miniwdl-lsf#7]
After the excellent and swift support with the last issue report, I thought we could discuss another potential enhancement.

### Problem: ###
Our LSF job scheduler is set up to submit jobs to one of several queues. The queue is chosen based on direct specification via the [`-q` argument](https://www.ibm.com/docs/en/spectrum-lsf/10.1.0?topic=o-q-1) or based on the runtime specified via the [`-W` argument](https://www.ibm.com/docs/en/spectrum-lsf/10.1.0?topic=o-w-1).
Not specifying either argument in the job submission leads to the assignment of the queue with the lowest runtime and premature termination of the jobs. I can solve this by specifying a queue in the [`extra_args`](https://github.com/adthrasher/miniwdl-lsf/blob/main/README.md?plain=1#L71-L72), but this implies using the same ("highest") queue for all jobs, irrespective of their actual requirements.

### Possible Solutions: ###
As far as I understand, WDL allows for the specification of [non-standard runtime attributes, called "hints"](https://github.com/openwdl/wdl/blob/main/versions/1.1/SPEC.md?plain=1#L1857C129-L1857C177). `__init__.py` includes a section demonstrating the [handling non-standard runtime attributes](https://github.com/adthrasher/miniwdl-lsf/blob/main/src/miniwdl_lsf/__init__.py#L76-L80) (introduced in the initial commit 4089dc973b04a0bf8cacdcf89a79da5d9f075ad0). This functionality could be used to handle specifications of task specific runtimes or queues.

I am unsure about the different handling of runtime attributes and hints. I could create a pull request, based on the code present in  `__init__.py`, but I think your input would be highly valuable here.

Thank you for your time.

### Versions: ###
```
$ miniwdl --version
miniwdl v1.11.1
miniwdl.plugin.container_backend        lsf_singularity = miniwdl_lsf:LSFSingularity    miniwdl-lsf 0.1.2
```

**Comment by adthrasher** [comment]
Thanks @Marc-Ruebsam for the suggestion! I have not, personally, used `hints` previously. The implementation has been in a state of flux in the WDL spec, but I don't think that prevents us from using them now. In WDL <= 1.1.1, `hints` are extra keys within the `runtime` section. WDL >= 1.2 currently plans to introduce a dedicated `hints` section (along with deprecating `runtime` in favor of `requirements`). 

That said, I'm initially opposed to supporting `-q` specifications at the task level. This seems like it would be highly cluster-specific and in opposition to the general portability goals of workflow languages. If you disagree, I'd like to hear your thoughts.

Supporting `-W` seems much more generalizable. As you noted, there is already some basic logic (with key `time_minutes) that is designed to handle that. At the moment, this information isn't passed to the LSF submission command, but that's a straightforward fix. 

Would support for `-W` sufficiently support your use case? Or is there more complexity needed?

Currently, there is a `shortTask` hint in the WDL spec that I haven't supported. I could potentially implement something using that hint (and possibly others, if generalizable). 

Can you share some additional details of your usage? For example, how do you divide tasks across queues? Is it a combination of memory, cpu, and runtime? Or are your queues more specific to a single resource requirement? Generally, how many "types" of jobs do you submit or how many queues do you want to submit to?

**Comment by Marc-Ruebsam** [comment]
Thank you for your reply, @adthrasher .

> That said, I'm initially opposed to supporting -q specifications at the task level. This seems like it would be highly cluster-specific and in opposition to the general portability goals of workflow languages. If you disagree, I'd like to hear your thoughts.
> 
> Supporting -W seems much more generalizable. As you noted, there is already some basic logic (with key `time_minutes) that is designed to handle that. At the moment, this information isn't passed to the LSF submission command, but that's a straightforward fix.
> 
> Would support for -W sufficiently support your use case? Or is there more complexity needed?

I don't know how widely used queues are. They are used on our cluster and are officially supported by LSF. I find them a neat way of distributing resources on the cluster and "roughly planning job submissions ahead". I understand your concerns in regard to portability, even though this would be a shortcoming of the corresponding workflows, not of miniwdl-lsf.
Nonetheless, I agree with you that support for -W is more straight forward and it will be sufficient for our cluster, as the runtime specification (referring to bsubs -W) will translate into a assignment of a specific queue.

> Currently, there is a shortTask hint in the WDL spec that I haven't supported. I could potentially implement something using that hint (and possibly others, if generalizable).

I was a bit disappointed to find little recommendations for WDLs hints or commonly used keys. Can you comment on the origin of the `time_minutes` and on the portability in this regard? I noticed the shortTask hint, but think a runtime specification (referring to bsubs -W) would be superior.

> Can you share some additional details of your usage? For example, how do you divide tasks across queues? Is it a combination of memory, cpu, and runtime? Or are your queues more specific to a single resource requirement? Generally, how many "types" of jobs do you submit or how many queues do you want to submit to?

I am but a humble user, so my knowledge of the sysadmin side is limited. To my knowledge, the cluster resources (i.e. nodes) are assigned to several queues. The different queues have maximum limits for runtime and memory. The memory requirement of the job is technically the only mandatory argument during job submission (to be set via -R 'rusage[mem=MEM]'). If no runtime is specified, the job will be submitted to the queue with the shortest runtime fulfilling the memory requirement. As this is often not desirable, a queue can be selected directly with the -q argument or by specifying the runtime with -W (the queue with the shortest runtime >= -W is selected). Additionally, the number of cores(/threads?) can specified with -n. Additional queue are available to handle nodes with specific hardware, e.g. GPU nodes.

**Comment by adthrasher** [comment]
I think queues, themselves, are broadly used. My issue here is that if you wrote something like

```
task foo {
   // --- snip ---
  runtime {
     queue: "my_queue"
 }
}
```
The queue name `my_queue` is likely to be specific to your cluster environment. So it would make the workflow less portable.

The `time_minutes` runtime `hint` was originally introduced by [`miniwdl-slurm`](https://github.com/miniwdl-ext/miniwdl-slurm) on which I based `miniwdl-lsf`. I intended to remove that functionality and did so in the submission, but left the linked artifact. I've looked through other `miniwdl` plugins as well as documentation for CWL and Nextflow. I've also had a conversation on the openWDL Slack channel. There hasn't been a consensus on general `hints` such as this one. 

Therefore I have opted to take the approach of adding an `lsf`-specific runtime hints section as suggested in the WDL 1.1 and 1.1.1 specifications. Currently, the only supported key will be `time`. The `time` hint is specified as an integer in minutes (though I have chosen the generic `time` so that in the future, it may accept other formats). 

I've written this in more detail in the updated README, but just to note here, the `hints` format will be changing in future releases of `WDL`.
